src/mcmini_private.cpp

```diff
@@ -200,59 +212,49 @@ mc_prepare_to_model_check_new_program()
                                            initialTransition);
 }
 
-MC_PROGRAM_TYPE
-mc_run_initial_trace()
+// nextBranchPoint is an out parameter.
+void
+mc_explore_branch(MCOptional<int> *nextBranchPoint)
 {
-  MC_PROGRAM_TYPE program = mc_fork_new_trace_at_main();
-  if (MC_IS_TARGET_PROGRAM(program)) return MC_TARGET_PROGRAM;
+  tid_t backtrackThread;
+
+  if (nextBranchPoint == nullptr /* initial branch */) {
+    mc_fork_new_trace();
+    backtrackThread = TID_MAIN_THREAD;
+  } else { // else next branch
+    const int bp = nextBranchPoint->unwrapped();
+    auto *sNext = &(programState->getStateItemAtIndex(bp));
+    backtrackThread = sNext->popThreadToBacktrackOn();
+
+    // Prepare the scheduler's model of the next trace
+    programState->reflectStateAtTransitionIndex(bp - 1);
+
+    mc_fork_next_trace_at_current_state();
+  }
 
-  mc_search_dpor_branch_with_initial_thread(TID_MAIN_THREAD);
+  mc_search_dpor_branch_with_thread(backtrackThread);
   mc_exit_with_trace_if_necessary(traceId);
-  program = mc_rerun_current_trace_as_needed();
+  mc_run_next_trace_for_debugger();
+
   traceId++;
-  if (MC_IS_TARGET_PROGRAM(program)) return MC_TARGET_PROGRAM;
-  return MC_SCHEDULER;
+  if (nextBranchPoint != nullptr) { // if this is not the initial branch
+    *nextBranchPoint = programState->getDeepestDPORBranchPoint();
+  }
 }
 
-MC_PROGRAM_TYPE
+void
 mc_do_model_checking()
 {
   mc_prepare_to_model_check_new_program();
 
-  // TODO: This idiom is fairly common... This simply allows
-  // the forked process to exit the constructor all of the stack
-  // frames find themselves in. Perhaps we could jump to the
-  // appropriate point with a getcontext()/setcontext() that will
-  // bring the trace process to the correct process as an alternative?
-  // It hurts readability to have the forked traces needing to escape
-  // in this way.
-  MC_PROGRAM_TYPE program = mc_run_initial_trace();
-  if (MC_IS_TARGET_PROGRAM(program)) return MC_TARGET_PROGRAM;
+  mc_explore_branch(nullptr); // nullptr means this is the initial branch
```

Comment:
Here you would then pass `MCOptional<int>::nil()` instead of `nullptr`

Reply:
And here, also, I would like to keep the code simple and easy to read.  Your change would make the code slightly slower.  And everyone will understand the current code, while the suggested code will appeal primarily to C++-enthusiasts.  I very much doubt that @lukagolf and @aayushi363 are in the habit of using these types of idioms.

Is there a reason to make this change, other than encouraging the reader of the code to go back and learn more about C++-specific idioms?

Reply:
As I mentioned above, the concept of an optional value should be familiar. All we're trying to do is add that into C++11 since `std::optional<T>` doesn't exist until later. In terms of speed, sure it could be slower sure. Again, if we're looking to move into a `C`-like codebase, we'd need to make bigger changes ultimately